### PR TITLE
AddMM

### PR DESCRIPTION
**Pull request by junha-l**
`AddMM` similar with [pytorch's addmm](https://pytorch.org/docs/stable/generated/torch.addmm.html)

It computes the generalized matrix multiplication as follow: 
```
output  = alpha * mat1 @ mat2 + beta * input
```
where, `alpha`, `beta` are scalar values, `mat1`, `mat2`, and `input` are tensors.

`mat1` and `mat2` should be 2D matrices with shape (m, n), (n,k), 
and `input` should have a broadcastable shape with (m, k).

TODO:
- [x] transpose as function argument
- [x] broadcast add
- [x] pybind
- [x] C++ unit test
- [x] python unit test

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/4184)

### CODE REVIEW

File: cpp/tests/core/Linalg.cpp [hunk 1]
```diff
@@ -89,6 +90,79 @@ TEST_P(LinalgPermuteDevices, Matmul) {
...
+    for (int i = 0; i < 8; ++i) {
+        EXPECT_TRUE(std::abs(C_data[i] - C_gt[i]) < EPSILON);
+    }
+
+    // Transposed addmm + alpha = -2.0 & beta = 5.0.
+    C = core::Tensor::Ones({2, 4}, dtype, device);
+    B_T_T = B_T.T();
+    core::AddMM(A, B_T_T, C, -2.0, 5.0);
+    EXPECT_EQ(C.GetShape(), core::SizeVector({2, 4}));
+    C_data = C.ToFlatVector<float>();
```
Comment (reyanshsolis) [inline cpp/tests/core/Linalg.cpp:140]: modify test: use Init to initialize gt tensor, and use `EXPECT_TRUE T.AllClose(T_gt, EPSILON)`.
Same for other tests.

File: cpp/pybind/core/linalg.cpp [hunk 2]
```diff
@@ -50,7 +51,18 @@ void pybind_core_linalg(py::module &m) {
...
+    m.def(
+            "addmm",
+            [](const Tensor &input, const Tensor &A, const Tensor &B,
+               double alpha, double beta) {
+                Tensor output =
+                        input.Expand({A.GetShape(0), B.GetShape(1)}).Clone();
+                AddMM(A, B, output, alpha, beta);
+                return output;
+            },
+            "Function to perform addmm of two 2D tensors with compatible "
```
Comment (reyanshsolis) [inline cpp/pybind/core/linalg.cpp:63]: mention the operation performed. ref. https://pytorch.org/docs/stable/generated/torch.addmm.html

File: cpp/open3d/core/linalg/AddMM.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include "open3d/core/Tensor.h"
+
+namespace open3d {
+namespace core {
+
+/// Computes matrix multiplication C = AB.
```
Comment (reyanshsolis) [inline cpp/open3d/core/linalg/AddMM.h:34]: may add `alpha` and `beta` also in the equation.

Comment (reyanshsolis) [reply cpp/open3d/core/linalg/AddMM.h:34]: also add detail docstring, like https://pytorch.org/docs/stable/generated/torch.addmm.html

File: cpp/open3d/core/linalg/AddMM.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+              int64_t n,
+              double alpha,
+              double beta,
+              bool gemmTrA,
+              bool gemmTrB,
+              int lda,
+              int ldb,
+              int ldc,
+              Dtype dtype);
+}  // namespace core
```
Comment (reyanshsolis) [inline cpp/open3d/core/linalg/AddMM.h:69]: nit: add one line

**Review by yxlao — CHANGES_REQUESTED** [review]

File: cpp/open3d/core/linalg/AddMM.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+#pragma once
+
+#include "open3d/core/Tensor.h"
+
+namespace open3d {
+namespace core {
+
+/// Computes matrix multiplication C = AB.
+void AddMM(
+        const Tensor& A, const Tensor& B, Tensor& C, double alpha, double beta);
```
[the remark this answers is not in the document]
Comment (junha-l) [reply cpp/open3d/core/linalg/AddMM.h:40]: Currently, we do not have a use case for this.
We can address this in later PR

**Review by yxlao — APPROVED** [review]

**Pull request merged**